fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -1,84 +1,140 @@
-            SupportedKeyClaim::Sub.to_string(),
-        ),
-        "18523124550523841778801820019979000409432455608728354507022210389496924497355",
-    );
-    map.insert(
-        (
-            OAuthProvider::Google.get_config().0,
-            SupportedKeyClaim::Email.to_string(),
-        ),
-        "",
-    );
-    map.insert(
-        (
-            OAuthProvider::Twitch.get_config().0,
-            SupportedKeyClaim::Sub.to_string(),
-        ),
-        "",
-    );
-    map.insert(
-        (
-            OAuthProvider::Twitch.get_config().0,
-            SupportedKeyClaim::Email.to_string(),
-        ),
-        "",
-    );
-    map
-});
+const MAX_HEADER_LEN: u16 = 500;
+const PACK_WIDTH: u16 = 248;
+const ISS: &str = "iss";
+const AUD: &str = "aud";
+const DUMMY_CLAIM: &str = "dummy_claim";
+const NUM_EXTRACTABLE_STRINGS: u8 = 5;
+const MAX_EXTRACTABLE_STR_LEN: u16 = 150;
+const MAX_EXTRACTABLE_STR_LEN_B64: u16 = 4 * (1 + MAX_EXTRACTABLE_STR_LEN / 3);
 
 /// Supported OAuth providers. Must contain "openid" in "scopes_supported"
 /// and "public" for "subject_types_supported" instead of "pairwise".
-#[derive(Debug)]
+#[derive(Debug, Hash, PartialEq, Eq)]
 pub enum OAuthProvider {
     /// See https://accounts.google.com/.well-known/openid-configuration
     Google,
     /// See https://id.twitch.tv/oauth2/.well-known/openid-configuration
     Twitch,
 }
 
+/// Struct that contains all the OAuth provider information. A list of them can
+/// be retrieved from the JWK endpoint (e.g. <https://www.googleapis.com/oauth2/v3/certs>)
+/// and published on the bulletin along with a trusted party's signature.
+// #[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OAuthProviderContent {
+    /// Key type parameter, https://datatracker.ietf.org/doc/html/rfc7517#section-4.1
+    pub kty: String,
+    /// RSA public exponent, https://datatracker.ietf.org/doc/html/rfc7517#section-9.3
+    pub e: String,
+    /// RSA modulus, https://datatracker.ietf.org/doc/html/rfc7517#section-9.3
+    pub n: String,
+    /// Algorithm parameter, https://datatracker.ietf.org/doc/html/rfc7517#section-4.4
+    pub alg: String,
+    /// kid
+    kid: String,
+}
+
+/// Reader struct to parse all fields.
+// #[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct OAuthProviderContentReader {
+    e: String,
+    n: String,
+    #[serde(rename = "use")]
+    my_use: String,
+    kid: String,
+    kty: String,
+    alg: String,
+}
+
+impl OAuthProviderContent {
+    /// Get the kid string.
+    pub fn kid(&self) -> &str {
+        &self.kid
+    }
+
+    /// Parse OAuthProviderContent from the reader struct.
+    pub fn from_reader(reader: OAuthProviderContentReader) -> FastCryptoResult<Self> {
+        if reader.alg != "RS256" || reader.my_use != "sig" || reader.kty != "RSA" {
```

Comment:
I'd like us to be careful with this check. After George's exploration of JWTs is done, let's see if all the headers actually contain these fields..

Reply:
happy to remove when new providers appear to not have this. for now, the three providers we want to launch with all satisfy this.